### ISSUE TITLE
Phone's control bar white/light in dark mode

### ISSUE THREAD

**Issue by tomas-hartman** [ankidroid/Anki-Android#5409]
###### Issue 

Hi, I encountered an issue with the way the control bar (where the phone control buttons are located) is displayed in night mode on my Samsung phone (android 9/one ui). In most other apps, color of the bar is black, however in AnkiDroid, its color is displayed as white/light which creates a really annoying contrast with the black background of the app itself, especially (but not only) during night time. Except for the uncomfortable conditions it makes, it can also have some negative effect on battery life on AMOLED displays. 

###### Screenshots

[black bar in other apps](https://user-images.githubusercontent.com/44980733/63222448-35099e00-c1a8-11e9-94e5-191758854b79.jpg)
[light bar in anki @ night mode](https://user-images.githubusercontent.com/44980733/63222449-35a23480-c1a8-11e9-9745-51dac457ed01.jpg)

###### Research
*Enter an [ x ] character to confirm the points below:*

[ x ] I have read the [support page](https://ankidroid.org/docs/help.html) and am reporting a bug or enhancement request specific to AnkiDroid

[ x ] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue

[ x ] I have searched for similar existing issues here and on the user forum

**Comment by mikehardy** [comment]
Ha - I see you already logged an issue. Is this with the latest beta?

**Comment by tomas-hartman** [comment]
Hi, nope it's the production version from Google Play. I'll join beta program and let you know. :)

**Comment by tomas-hartman** [comment]
OK, still the same with 2.9beta5.

**Comment by mikehardy** [comment]
Bummer - but thank you for testing it

**Comment by timrae** [comment]
Please post a screenshot

**Comment by timrae** [comment]
Ah nevermind, I see

**Comment by Kartikkman** [comment]
Hii @mikehardy , 

I want to contribute towards this project . I am a beginner in this regards . 
Should I proceed with this issue ?? 

Thank You

**Comment by mikehardy** [comment]
Sure @Kartikkman - and welcome! Check our [wiki for info on Development](https://github.com/ankidroid/Anki-Android/wiki/Development-Guide) (and note that apparently you need to turn Instant Run off in Android Studio per #5406) 

That should get you all set up so that you compile the current code and having it run on an emulator on your machine, at that point you're ready to start making changes

This change specifically? I am *guessing* (could be wrong) that there will be some sort of XML style attribute we can set? Probably something missing in https://github.com/ankidroid/Anki-Android/blob/master/AnkiDroid/src/main/res/values/theme_black.xml or https://github.com/ankidroid/Anki-Android/blob/master/AnkiDroid/src/main/res/values/theme_dark.xml - but I could be totally wrong there

Good luck :-)

**Comment by Kartikkman** [comment]
Hii @mikehardy  , I have setup Git Stuff ( forking , .. cloning & all ) . Code is now set up & is running on my phone .

**Comment by Kartikkman** [comment]
Hey @mikehardy , Issue is a simple Fix . Setting the " navigationBarColor" Property in [https://github.com/ankidroid/AnkiAndroid/blob/master/AnkiDroid/src/main/res/values/theme_black.xml]()

`<item name="android:navigationBarColor">@color/black</item>`

Screen Shots : 

**Before :** 

![ScreenShot_Before](https://user-images.githubusercontent.com/28975022/63477931-2cb1ab80-c4a5-11e9-9bcc-d6b899f7e09e.png)

**After :** 

![ScrrenShot_After](https://user-images.githubusercontent.com/28975022/63477935-320ef600-c4a5-11e9-8ebd-799b913c269b.png)


But This was introduced after API Level 21 , so won't support on KitKat . 

Another way , I think we can proceed with this is use the method which works on API level less than 21 . Below is the method : 

`window.setNavigationBarColor(@ColorInt int color)`

We can call this method when "Nighmode" Switch is activated & dectivated  . Setting Navigation Bar to : 

 Black Color when Activated 
White Color when Deactivated 

@mikehardy , Your views on it ?? 

Kartik Sikka

**Comment by mikehardy** [comment]
Hey cool - that was quick. I'd go with a solution that was clean first - and the XML is overlayed for API, so you can change it here and it won't crash old phones: https://github.com/ankidroid/Anki-Android/blob/master/AnkiDroid/src/main/res/values-v21/styles.xml

Then for APIs less than 21...Google tells me that's a total of 35,000 users out of 1,200,000, with about 10,000 about to be below minimum support shortly, and the rest declining at 15,000 a year. I'd save the effort as doing API-specific code things isn't great and there are more important things if you've got a dev environment set up. For instance #5097 this one is hurting a lot of people and we haven't even reproduced it yet :-(

So if the style overlay from values-v21 worked I'd PR that and go with it

**Comment by Kartikkman** [comment]
Hey @mikehardy 

I have made changes in [https://github.com/ankidroid/Anki-Android/blob/master/AnkiDroid/src/main/res/values-v21/styles.xml](https://github.com/ankidroid/Anki-Android/blob/master/AnkiDroid/src/main/res/values-v21/styles.xml) 

It is **Working**  . Below are the screenshots  : 

[Working](https://user-images.githubusercontent.com/28975022/63479472-122f0080-c4ac-11e9-9ad0-c43afbb2c80a.png)

But there is flaw with above thing , It doesn't work while we are in Settings Window ( I think we call it as Preference Fragment  ??) . Below is the Screenshot  : 

[NoChangeInSettingsWindow](https://user-images.githubusercontent.com/28975022/63479615-91bccf80-c4ac-11e9-815c-f6c72185789c.png)

@mikehardy  , Are we Okay with this ?? Or Should I look for the solution of above sub problem as well ??

**Comment by mikehardy** [comment]
Part of the fun is sort of digging in and searching around on stackoverflow when you have no idea what is going on ;-). The things you learn...

For the preferences not respecting navigationBarColor - that's really weird. It seems incomplete without that - digging in on that would be nice - but if there is no good answer then at least it is rarely used vs the navigationBar in the main theme

For #5097 - my first step would be to try to find a report that was as close to an emulator as I had (same Android version) and configure it for their language, clean everything out, do an install and see if you can reproduce it based on the steps the users report. We haven't found out what's causing it at all.

**Comment by mikehardy** [comment]
I see the "Layout Editor" tool in Android Studio, that I'd never played with before. Maybe that will allow inspection of the navigation bar while your emulator is showing the odd preferences behavior? I literally just tried it the first time out of curiosity so I have no good specifics on it, and the Pixel 3 / API29 emulator I tried doesn't exhibit the behavior so all I have is the idea of trying the tool. Hope it helps?

**Comment by Kartikkman** [comment]
Hii @mikehardy  , Sorry for Replying Late ( was busy with Office Stuff ) . 
Well for the above I need to learn " Layout Inspector " , I am trying with it !!!

**Comment by Kartikkman** [comment]
@mikehardy Are we still working on it ??

**Comment by mikehardy** [comment]
sorry - if you edit a previous comment instead of making a new one, it does not trigger any sort of notification for me, so I was simply unaware you were looking for feedback

Seems that we don't have a good answer for preferences - I don't have the answers to your questions, I'm not great at Android layout, sorry

If there's not 100% answer it might be time to at least make a pull request - have you ever made one?

Just in case you haven't (ignore me if you have) - typically the work you've been doing would be on a git branch (to leave master clean) - checked out like so `git checkout -b 'issue5409/fix-status-bar-dark-mode'` or whatever you want to call the branch

Once you have it working well locally, make sure `./gradlew jacocoUnitTestReport` still works in your local environment, then `git push` back up to your branch on your fork on github

At that point in the github.com web interface it will probably prompt you to make a pull request, do that and just follow the template. Should be easy (it's never easy the first time)? :-)

**Comment by Kartikkman** [comment]
Hii @mikehardy , I tried to create Pull Request , Can You Please Check it out ?? 

One More Thing : 

This gradlew Test stuck at 99 % .
![gradlew Test ](https://user-images.githubusercontent.com/28975022/63656099-b87d4d80-c7ad-11e9-8085-e2638576061b.png)

@mikehardy , I am thankful for all your support , I am learning this all & will do my best to Deliver back to you . 

Can you assign me next issue  ( for Beginners ) , you may suggest me ?? 

Thanks Again !!!

**Comment by mikehardy** [comment]
Huh, that's strange regarding the 99% completion. Honestly I haven't tried it on windows ever, I wonder if there is an operating system specific error in the tests somewhere? I run them on ubuntu and macOS. I'll have to check it. 

The PR looks fine though - and I don't see how it could do anything that would cause a problem for the tests. I'm sure it will pass our continuous integration tests

Thanks for the effort!

I think #5097 is our highest priority issue by far. It may not be easy but just trying to reproduce it at least is really valuable effort.

As for another good first issue I think #4968 is a good one - it gets into the Java code but is not a difficult change so it makes for a good start

**Comment by pablode** [comment]
@Kartikkman to answer your question, the preferences theme is defined in the [App Manifest](https://github.com/ankidroid/Anki-Android/blob/7ca6710fbdc1ed51982e5ba314ed74a49de0681e/AnkiDroid/src/main/AndroidManifest.xml#L214).
```xml
<activity
    android:name="com.ichi2.anki.Preferences"
    android:exported="false"
    android:configChanges="keyboardHidden|orientation|locale|screenSize"
    android:label="@string/preferences_title"
    android:theme="@style/LegacyActionBarLight"/>
```
However, changing it won't fix the issue, because [it gets swapped at runtime](https://github.com/ankidroid/Anki-Android/blob/7ca6710fbdc1ed51982e5ba314ed74a49de0681e/AnkiDroid/src/main/java/com/ichi2/anki/Preferences.java#L110):
```java
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        Themes.setThemeLegacy(this);
        // ...
    }
```
If you follow the [code trail](https://github.com/ankidroid/Anki-Android/blob/7ca6710fbdc1ed51982e5ba314ed74a49de0681e/AnkiDroid/src/main/java/com/ichi2/themes/Themes.java#L67) you will find that neither `Theme_Dark_Compat` nor `Theme_Black_Compat` are being used. Instead, `LegacyActionBarDark` or `LegacyActionBarBlack` is chosen. You have to add the `navigationBarColor` attribute to them, too.

**Comment by mikehardy** [comment]
@pablode nice detective work - this could be a quick win with just another tiny XML resource tweak @Kartikkman

**Comment by Kartikkman** [comment]
Hey @pablode  that's amazing , This is something I could never have come upto . I am making the change to Code Now  , ( Would you like to share your Secret Recipe: How you find out this ) ?

**Comment by pablode** [comment]
There is no secret recipe, it's just years of accumulated knowledge and some basic deductive reasoning. Invest some time and eventually, you will get there, too!

**Comment by mikehardy** [comment]
Sounds about right re: experience. It just takes time. I'm on maybe 1.5 years of Android work and I couldn't find the preference theme either when I first trawled through. I must need more time :-)

**Comment by david-allison-1** [comment]
I believe this is closed in #5417, please reopen if this is not the case.
